### PR TITLE
test: add missing integration markers, restore reloaded vault modules, track dead skip (#497)

### LINKED ISSUE #497: tests: missing integration markers, vault SDK reload poisoning, untracked dead skip
## Summary

Three test-hygiene defects undermine the lane contract the suite is built on: unmarked binary-backed tests leak into
the unit lane (and the tests/scanner real-binary classes run in NO CI lane at all), four vault test modules
`importlib.reload()` package modules under mock SDKs without restoring (the exact pattern CLAUDE.md forbids, leaving
a vault client that cannot fail for the rest of the process), and the repo's only unconditional skip suppresses its
only non-default-region AWS test with no tracking issue. Every item below was reproduced twice (original reviewer +
independent verifier) on origin/main commit cb68c97.

## Findings

- [ ] **tests/integration/test_kingfisher_scanner.py and test_ephemeral_keys.py carry no `integration` marker — 16
  tests leak into the `-m "not integration"` unit lane, and the tests/scanner real-binary Integration classes run in
  no CI lane at all.** `tests/integration/test_kingfisher_scanner.py:17`
  **Repro:** `uv run pytest tests/integration/test_ephemeral_keys.py tests/integration/test_kingfisher_scanner.py
  -m "not integration" --collect-only` → 16 collected; unit-run `--durations` shows kingfisher real scans as the 11
  slowest "unit" tests (~23s).
  **Expected:** every file under tests/integration/ carries `pytest.mark.integration` (+ backend marker) so the
  `-m` selectors decide lanes, not directory luck.
  **Actual:** kingfisher binary tests run in the unit lane and never under `-m integration`; the
  gitleaks/trufflehog/trivy/infisical/talisman Integration classes in tests/scanner are skipped in the unit lane
  (binary absent), deselected from `-m integration` (unmarked), and not collected by integration-tests.yml
  (path-scoped).
  **Verifier note:** talisman/trivy/infisical do get real CI coverage via test_scanner_external_binaries.py;
  gitleaks/trufflehog/kingfisher get none. Distinct from #453 (missing `aws` marker).

- [ ] **Unit vault tests `importlib.reload()` envdrift.vault.<provider> under MagicMock SDKs without restoring —
  after the unit lane, envdrift.vault.hashicorp is permanently poisoned in-process.**
  `tests/unit/test_vault_hashicorp.py:38`
  **Repro:** run test_cov_vault_hashicorp.py then test_vault_hashicorp.py in one process; afterwards
  `get_vault_client('hashicorp', url='http://localhost:9999').authenticate()` SUCCEEDS against a closed port
  (`_hvac` is a SimpleNamespace, `is_authenticated()` a MagicMock).
  **Expected:** fixtures restore the module (reload again in `finally`, as test_cov_vault_hashicorp.py:80-91 does) or
  avoid reload entirely; CLAUDE.md forbids exactly this pattern.
  **Actual:** four modules (hashicorp :38, azure :92, aws :82, gcp :116) leave mock SDKs bound for the rest of the
  process; the gcp contamination is ACTIVE in the standard lane (test_vault_gcp_paths.py runs after test_vault_gcp.py).
  **Verifier note:** masked only by alphabetical collection luck — test_vault_hashicorp.py is currently the last file
  in tests/unit/; any reordered run (`--lf`, xdist, a new file sorting after it) silently gets the fake client.

- [ ] **test_client_with_different_region is unconditionally `@pytest.mark.skip`'d ("LocalStack region handling is
  flaky in CI") with no issue reference — a permanently dead test.** `tests/integration/test_aws_integration.py:520`
  **Repro:** `grep -rn 'pytest.mark.skip(' tests/` → exactly one hit; the test skips even with a healthy local
  LocalStack while its sibling passes.
  **Expected:** fix the flake, scope the gate conditionally to CI, or reference a filed issue per the repo's
  skip/xfail convention.
  **Actual:** the only non-default-region AWS client test is dead on all environments; nothing tracks re-enabling it.

## Suggested fix direction

- Add `pytestmark = [pytest.mark.integration, ...]` to the two unmarked files and mark (or relocate) the five
  tests/scanner `Test*Integration` classes so each runs in exactly one lane.
- Adopt the restore pattern of test_cov_vault_hashicorp.py:80-91 (finally-reload after un-patching) in the four
  mock-SDK fixtures, or rewrite them in the reload-free style of test_vault_gcp_paths.py.
- Replace the bare skip at test_aws_integration.py:520 with a conditional gate or
  `xfail(strict=False, reason="... (see #NNN)")` against a filed issue.

_Part of the 2026-06 final deep-review (batch: TEST-3). Each item was independently reproduced and re-verified before filing._

### PR DESCRIPTION
**Pull request by jainal09**
## Summary

Resolves all three test-hygiene defects from #497: missing `integration` markers (16 tests leaking into
the unit lane and five real-binary scanner classes running in NO lane), mock-SDK `importlib.reload()`
poisoning of the four vault provider modules, and the repo's only unconditional dead skip. Ships
regression tests (`tests/unit/test_suite_hygiene.py`) that were written first and proven to fail on
unfixed code.

## Fixes

| Issue checklist item | Change |
|---|---|
| `test_kingfisher_scanner.py` / `test_ephemeral_keys.py` unmarked; 16 tests leak into the unit lane | Both files now carry `pytest.mark.integration` (kingfisher keeps its binary skip-gate) |
| `tests/scanner` real-binary `Test*Integration` classes run in no lane | `@pytest.mark.integration` added to the gitleaks/trufflehog/trivy/infisical/talisman classes; they are now selected by `-m integration` and skip-gate on the binary |
| Vault unit tests reload `envdrift.vault.<provider>` under MagicMock SDKs without restoring | All four fixtures (`hashicorp`, `aws`, `azure`, `gcp`) now finally-reload the provider module after `patch.dict` unwinds, re-binding the real SDK (per-provider modules only — `envdrift.vault` itself is never reloaded). The bare restore in `test_init_without_boto3_raises` is also now exception-safe |
| `test_client_with_different_region` permanently dead behind a bare skip with no issue reference | Skip removed and the test fixed: Secrets Manager namespaces secrets per region (verified against live LocalStack), so the old create-in-us-east-1/read-from-eu-west-1 flow could never pass. The test now creates the secret in eu-west-1, reads it back via `AWSSecretsManagerClient(region="eu-west-1")`, and asserts region isolation |
| (guard) | `docs/reference/testing.md` documents the suite-hygiene rules; a new hygiene test rejects any future unconditional `pytest.mark.skip` lacking a `#NNN` issue reference |

## Test evidence

New regression tests in `tests/unit/test_suite_hygiene.py`, written first and run against unfixed code:

| Test | Pre-fix failure (verbatim) | Post-fix |
|---|---|---|
| `test_integration_tree_is_excluded_from_unit_lane` | `AssertionError: 16 test(s) under tests/integration/ leak into the unit lane (missing pytest.mark.integration)` | passed |
| `test_scanner_binary_integration_classes_run_in_integration_lane` | `Scanner integration classes not selected by -m integration: ['TestGitleaksIntegration', 'TestTrufflehogIntegration', 'TestTrivyIntegration', 'TestInfisicalIntegration', 'TestTalismanIntegration']` | passed |
| `test_vault_unit_tests_restore_real_provider_modules` | `vault provider modules left poisoned after unit tests: hashicorp._hvac is namespace(... Client=<MagicMock ...>); aws._boto3 is <MagicMock ...>; azure._SecretClient is <MagicMock ...>; gcp._secretmanager is namespace(SecretManagerServiceClient=<MagicMock ...>)` | passed |
| `test_no_unconditional_skip_without_issue_reference` | `Unconditional pytest.mark.skip without a tracking-issue reference ... ['tests/integration/test_aws_integration.py:520']` | passed |

The restore checker runs the four vault unit-test files with the real pytest in a subprocess, then asserts
module bindings are real again — including the behavioral symptom from the issue: `authenticate()` against
a closed port must fail.

Gates (all in this worktree, live LocalStack/Vault/Lowkey-Vault backends):

- `ruff check` / `ruff format --check` / `pyrefly check` (0 errors) / `bandit` / `make lint-docs` — clean
- `pytest -m "not integration"` — 2508 passed (kingfisher/ephemeral now deselected)
- `pytest -m integration` — 340 passed, 14 skipped (incl. the re-enabled region test against real LocalStack; trivy/infisical/talisman integration classes ran with real binaries)
- full single-process suite (`uv run pytest`, the Publish-workflow order) — 2848 passed, 14 skipped
- hygiene tests re-run under `FORCE_COLOR=1` — passed

Fixes #497

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/jainal09/envdrift/pull/532?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Documentation**
  * Added clearer guidance for test suite hygiene and lane selection.

* **Bug Fixes**
  * Improved isolation between unit and integration tests.
  * Fixed several SDK-related test setups so mocked imports no longer affect later runs.
  * Tightened region handling checks to ensure secrets stay isolated by region.

* **Tests**
  * Marked integration tests consistently for the integration lane.
  * Added regression coverage for test selection, skip usage, and module restore behavior.
  * Updated external-binary test handling to skip cleanly when tools are unavailable.

<h3>Greptile Summary</h3>

This PR resolves three test-hygiene defects from #497: 16 integration tests leaking into the unit lane (missing `pytest.mark.integration` on `test_ephemeral_keys.py`, `test_kingfisher_scanner.py`, and five scanner `Test*Integration` classes), mock-SDK `importlib.reload()` in four vault fixtures leaving MagicMock SDKs in `sys.modules`, and one permanently-dead unconditional skip in the AWS integration suite.

- **Lane markers**: `pytest.mark.integration` added to `test_ephemeral_keys.py`, `test_kingfisher_scanner.py`, and the `Test*Integration` classes in all five scanner test files; they are now correctly selected by `-m integration` and excluded from `-m "not integration"`.
- **Vault fixture restoration**: All four vault mock-SDK fixtures (`hashicorp`, `aws`, `azure`, `gcp`) now wrap the `patch.dict` context in `try/finally` and call `importlib.reload()` in the `finally` block, ensuring the real SDK bindings are restored even when tests fail mid-assertion.
- **Dead skip removed**: `test_client_with_different_region` was rewritten to correctly test region isolation by creating the secret in eu-west-1 and asserting the us-east-1 fixture secret is not visible there; a new `tests/unit/test_suite_hygiene.py` ships regression tests (using AST scanning and subprocess-isolated pytest runs) that pin all four invariants against future regressions.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all changes are test-only, addressing genuine hygiene defects with no modifications to production source code.

Every change is confined to the test tree and documentation. The vault fixture restorations are mechanically correct (try/finally around a patch.dict context with a restoring reload after the context unwinds). The lane markers are additive and verified by the new regression tests. The rewritten AWS region-isolation test creates its own secret, asserts the correct value, and cleans up unconditionally. The AST-based skip scanner handles all edge cases proven by the bundled regression test. No logic changes touch envdrift production code.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| tests/unit/test_suite_hygiene.py | New regression test file enforcing four lane-contract invariants via subprocess and AST analysis; logic is correct and the AST-based skip scanner handles all edge cases (bare decorators, reason strings, skipif, pytest.param). |
| tests/integration/test_aws_integration.py | Removed unconditional dead skip; rewrote test to create an eu-west-1 secret, verify it's readable by an eu-west-1 client, and assert region isolation with a SecretNotFoundError for the us-east-1 fixture secret; cleanup in a finally block. |
| tests/unit/test_vault_hashicorp.py | Module-scoped autouse fixture wrapped in try/finally; importlib.reload in finally block restores real hvac bindings regardless of test failures. |
| tests/unit/test_vault_aws.py | mock_boto3 fixture and test_init_without_boto3_raises both now use try/finally for importlib.reload, ensuring module restoration is exception-safe. |
| tests/unit/test_vault_azure.py | mock_azure fixture refactored to try/finally with a restoring reload in finally; same correct pattern as the other vault fixture changes. |
| tests/unit/test_vault_gcp.py | mock_gcp fixture now wraps the patch.dict context in try/finally and reloads the module in finally to restore real GCP SDK bindings. |
| tests/integration/test_ephemeral_keys.py | Added pytestmark = pytest.mark.integration to place the file in the integration lane. |
| tests/integration/test_kingfisher_scanner.py | pytestmark changed from a bare skipif to a list combining pytest.mark.integration with the binary-presence skipif guard. |
| tests/scanner/test_gitleaks.py | Added @pytest.mark.integration to TestGitleaksIntegration class so it appears in the integration lane. |
| docs/reference/testing.md | Added Suite Hygiene Rules section documenting the four lane-contract invariants enforced by test_suite_hygiene.py. |

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/497-test-hy..."](https://github.com/jainal09/envdrift/commit/471e7b115efc8e1a46944abbf2721d0ee05b423a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37247934)</sub>

### CODE REVIEW

File: tests/unit/test_suite_hygiene.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    offenders: list[str] = []
+    for path in sorted(TESTS_DIR.rglob("*.py")):
+        if path.resolve() == this_file:
+            continue
+        text = path.read_text(encoding="utf-8")
+        for match in skip_re.finditer(text):
+            window = text[match.start() : match.start() + 300]
+            if not issue_ref_re.search(window):
+                line = text.count("\n", 0, match.start()) + 1
+                offenders.append(f"{path.relative_to(REPO_ROOT)}:{line}")
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_suite_hygiene.py]: Confirmed real (latent): a file containing only `# never use pytest.mark.skip` (no issue number within 300 chars) was flagged by the old text scan. Fixed in 812dc00ec378d0d79886ee54dffa29cc9c447584: the scan now parses each test file with `ast` and matches only real `pytest.mark.skip` attribute/call nodes (`skipif` still never matches); the tracking-issue reference must appear in one of the skip call's string arguments (e.g. `reason="... #NNN"`). Comments and docstrings can no longer match, which also let us drop the self-exclusion of the hygiene file. Added a regression test, `test_bare_skip_scan_matches_code_not_comments_or_docstrings`, covering comment/docstring mentions, a bare decorator, reasons with/without an issue ref, `skipif`, and `pytest.param(marks=pytest.mark.skip(...))`.

File: tests/unit/test_suite_hygiene.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                "tests/unit/test_vault_gcp.py",
+                "-q",
+                "--no-cov",
+                "-p",
+                "no:cacheprovider",
+            ]
+        )
+        if rc != 0:
+            print(f"inner pytest run failed with exit code {rc}")
+            sys.exit(2)
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_suite_hygiene.py:152]: This is deliberate and load-bearing, not a defect: the #497 contamination is only observable through a shared `sys.modules`. The Publish workflow runs the whole suite in one process (`uv run pytest`), which is exactly the environment where a leaked MagicMock SDK poisons later tests. If the inner run were subprocess-isolated, the checker process would always import fresh modules and this test could never fail — the check would be vacuous. The scenario described (a collection-time import leaving a partially-reloaded module that survives fixture teardown) is precisely the pollution this check must catch: it would equally poison later tests in a full-suite run, so a failure here is a true positive. Documented this rationale in the comment above `_VAULT_RESTORE_CHECKER_SCRIPT` in 812dc00ec378d0d79886ee54dffa29cc9c447584.

File: tests/unit/test_suite_hygiene.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        env=_subprocess_env(),
+        capture_output=True,
+        text=True,
+        timeout=300,
+    )
+
+    assert result.returncode == 0, (
+        f"vault module restore check failed (exit {result.returncode}):\n"
+        f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
+    )
```
Comment (jainal09) [inline tests/unit/test_suite_hygiene.py]: Fixed in 812dc00ec378d0d79886ee54dffa29cc9c447584: the embedded checker script is extracted to a module-level `_VAULT_RESTORE_CHECKER_SCRIPT` constant, so `test_vault_unit_tests_restore_real_provider_modules` is now ~27 lines (docstring included), well under the 70-line threshold.

File: tests/unit/test_suite_hygiene.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        for match in skip_re.finditer(text):
+            window = text[match.start() : match.start() + 300]
+            if not issue_ref_re.search(window):
+                line = text.count("\n", 0, match.start()) + 1
+                offenders.append(f"{path.relative_to(REPO_ROOT)}:{line}")
+
+    assert offenders == [], (
+        "Unconditional pytest.mark.skip without a tracking-issue reference "
+        f"(fix the test or cite the issue in the reason): {offenders}"
+    )
```
Comment (jainal09) [inline tests/unit/test_suite_hygiene.py:327]: Fixed in 812dc00ec378d0d79886ee54dffa29cc9c447584: the skip scan was rewritten AST-based (see the Greptile thread about regex false positives), and `test_no_unconditional_skip_without_issue_reference` is now a flat comprehension plus a single assert. The matching logic lives in small helpers (`_is_pytest_mark_skip`, `_skip_call_references_issue`, `_bare_skip_offenders`) built from guard-free comprehensions, so no function has nested conditional blocks.

**Comment by jainal09** [comment]
@coderabbitai full review
